packages/base/source/adt/union/union.js

```diff
@@ -82,8 +108,15 @@ instead to check if a value belongs to the ADT variant.`);
        *   where 'b = 'a[`@@folktale:adt:tag]
        */
       matchWith(pattern) {
-        return pattern[name](this);
-      } 
+        assertObject(`${name}#matchWith`, pattern);
```

Comment:
We can change:

```js
assertObject(`${name}#matchWith`, pattern)
```

To:

```js
assertObject(`${typeId}'s ${name}#matchWith`, pattern);
```

To output the Union type as well, which should help understanding what kind of thing you've got. Though that's about the most we can do without using something like TypeScript.